File: AdventOfCode.2024/Solutions/Day01_Part2.cs
namespace AdventOfCode.Solutions;

internal sealed class Day01_Part2 : PuzzleSolution
{
    private const string DAY = "01";
    public static string FileName { get; } = $"Data/{DAY}.input";
    public static string TestFileName { get; } = $"Data/{DAY}.sample";
    public static string TestOutputExpected { get; } = "31";

    public static string Solve(StreamReader reader)
    {
        var leftValues = new List<int>();
        var rightValues = new Dictionary<int, int>();
        while (reader.ReadLine() is { } line)
        {
            var firstspace = line.IndexOf(' ');
            var lastspace = line.LastIndexOf(' ');
            var left = int.Parse(line[..firstspace]);
            var right = int.Parse(line[(lastspace + 1)..]);
            leftValues.Add(left);
            if (rightValues.TryGetValue(right, out int value))
            {
                rightValues[right] = ++value;
            }
            else
            {
                rightValues[right] = 1;
            }
        }

        var similarity = 0;

        foreach (var left in leftValues)
        {
            if (rightValues.TryGetValue(left, out int value))
            {
                similarity += value * left;
            }
        }

        return similarity.ToString();
    }
}

File: javascript/2024/solutions/day01part2.js
export const inputFile = '01.input';
export const testFile = '01.sample';
export const testOutputExpected = "31";

export function solve(input) {
    let left = [];
    let right = new Map();

    input.split('\n').forEach(line => {
        let parts = line.split(/\s+/);
        left.push(parseInt(parts[0]));
        let rightValue = parseInt(parts[1]);
        if (right.has(rightValue)) {
            right.set(rightValue, right.get(rightValue) + 1);
        } else {
            right.set(rightValue, 1);
        }
    });

    var similarity = 0;
    for (let i = 0; i < left.length; i++) {
        var leftValue = left[i];
        if (right.has(leftValue)) {
            similarity += right.get(leftValue) * leftValue;
        }
    }
    
    return similarity.toString();
}
